substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JmxServerFeature.java

```diff
@@ -0,0 +1,264 @@
+            RuntimeClassInitializationSupport rci = ImageSingletons.lookup(RuntimeClassInitializationSupport.class);
+            rci.initializeAtBuildTime("sun.rmi.server.Activation$ActivationSystemImpl_Stub", "JDK11 JMX support");
+        }
+    }
+
+    private static void handleNativeLibraries(BeforeAnalysisAccess access) {
+
+        // This is required for password authentication. JMX checks the restrictions on the password file via a JNI native method.
+        NativeLibrarySupport.singleton().preregisterUninitializedBuiltinLibrary("management_agent");
+        BeforeAnalysisAccessImpl beforeAnalysisAccess = (BeforeAnalysisAccessImpl) access;
+        beforeAnalysisAccess.getNativeLibraries().addStaticJniLibrary("management_agent");
+        // Resolve calls to jdk_internal_agent* as builtIn. For calls to native method isAccessUserOnly0.
+        PlatformNativeLibrarySupport.singleton().addBuiltinPkgNativePrefix("jdk_internal_agent");
+    }
+
+    @Override
+    public void beforeAnalysis(BeforeAnalysisAccess access) {
+        handleNativeLibraries(access);
+        registerJMXAgentResources();
+        configureJNI();
+        configureSerialization(access);
+        configureReflection(access);
+        configureProxy(access);
+    }
+
+    private static void registerJMXAgentResources() {
+        ResourcesRegistry resourcesRegistry = ImageSingletons.lookup(ResourcesRegistry.class);
+
+        resourcesRegistry.addResourceBundles(ConfigurationCondition.alwaysTrue(),
+                "jdk.internal.agent.resources.agent");
+
+        resourcesRegistry.addResourceBundles(ConfigurationCondition.alwaysTrue(),
+                "sun.security.util.Resources"); // required for password auth
+    }
+
+    private static void configureProxy(BeforeAnalysisAccess access) {
+        DynamicProxyRegistry dynamicProxySupport = ImageSingletons.lookup(DynamicProxyRegistry.class);
+
+        dynamicProxySupport.addProxyClass(access.findClassByName("java.rmi.Remote"),
+                access.findClassByName("java.rmi.registry.Registry"));
+
+        dynamicProxySupport.addProxyClass(access.findClassByName("javax.management.remote.rmi.RMIServer"));
+
+        dynamicProxySupport.addProxyClass(access.findClassByName("jdk.management.jfr.FlightRecorderMXBean"),
+                access.findClassByName("javax.management.NotificationEmitter"));
+    }
+
+    private static void configureJNI() {
+        try {
+            JNIRuntimeAccess.register(Arrays.class);
+            JNIRuntimeAccess.register(Arrays.class.getDeclaredMethod("asList", Object[].class));
+        } catch (Exception e) {
+            throw VMError.shouldNotReachHere("ManagementFeature JNI Configuration failed: " + e);
+        }
+    }
+
+    private static void configureSerialization(BeforeAnalysisAccess access) {
+        String[] classes = new String[] {
+                "[B", "com.oracle.svm.core.jdk.UnsupportedFeatureError",
```

Comment:
Yes it is JDK specific

Reply:
If a feature requires that much configuration, it is better to use config-*.json files. Combined with https://www.graalvm.org/22.0/reference-manual/native-image/Reflection/#conditional-configuration it should be possible to make sure that the config data is only applied if JMX support is requested.

Reply:
> If a feature requires that much configuration, it is better to use config-*.json files.

I just realized that unfortunately that does not work in this case because this is not an external Feature thus the `META-INF/native-image` based approach cannot be used. Sorry for the noise!

Reply:
At the moment I don't see a better way to do this. For the build-time init future programming model changes will help in the future. For reachability metadata, we would need both programming model changes and the JDK modifications.